### ISSUE TITLE
invalid iptables / exit node not working

### ISSUE THREAD

**Issue by neyrond** [Fluent-networks/tailscale-mikrotik#30]
Launched tailscale container (fluent-networks/tailscale-mikrotik:latest) on hap ac3. It looks it's not working as an exit node. 

`/ # tailscale status` inside the container gives this error: 

```
# Health check:
#     - adding [-i tailscale0 -j MARK --set-mark 0x40000/0xff0000] in v4/filter/ts-forward: running [/sbin/iptables -t filter -A ts-forward -i tailscale0 -j MARK --set-mark 0x40000
/0xff0000 --wait]: exit status 1: iptables: Invalid argument. Run `dmesg' for more information.
```

and
```
/ # dmesg
dmesg: klogctl: Operation not permitted

```

**Comment by frankedwards** [comment]
Hi

This appears to be an issue with Alpine 3.18 + iptables 1.8.9. Updating to Alpine 3.19 + iptables-1.8.10 appears to allow exit node functionality to be used.

Could you please test using 0.1.33 / latest.

**Comment by neyrond** [comment]
It's working now, thank you!

**Comment by Devinaille** [comment]
Same issue, and I upgrade to 0.1.33, there is a new problem:
```
# Health check:
#     - setting up filter/ts-input: running [/sbin/iptables -t filter -N ts-input --wait]: exit status 4: iptables v1.8.10 (nf_tables): Could not fetch rule set generation id: Invalid argument
```

**Comment by DusKing1** [comment]
> Same issue, and I upgrade to 0.1.33, there is a new problem:
> 
> ```
> # Health check:
> #     - setting up filter/ts-input: running [/sbin/iptables -t filter -N ts-input --wait]: exit status 4: iptables v1.8.10 (nf_tables): Could not fetch rule set generation id: Invalid argument
> ```

I have same issue on my end, do you have any solution for this?

**Comment by frankedwards** [comment]
Please test using 0.1.34.
```
0.1.34
Migrate to nftables.
```

**Comment by RomRider** [comment]
Same issue with nftables.
```
# Health check:
#     - setting up filter/ts-input: running [/sbin/iptables -t filter -N ts-input --wait]: exit status 4: iptables v1.8.10 (nf_tables): Could not fetch rule set generation id: Invalid argument
```

**Comment by frankedwards** [comment]
Hi,

Can you check TS_DEBUG_FIREWALL_MODE is being set in /usr/local/bin/tailscale.sh

```
mikrotik:~# cat /usr/local/bin/tailscale.sh | grep TS_DEBUG_FIREWALL_MODE
TS_DEBUG_FIREWALL_MODE=nftables
```

**Comment by RomRider** [comment]
Yes, but it needs `export` to work (I modified the script).

Result being:
```
# Health check:
#     - create table: get table: get tables: netlink receive: invalid argument
```

It seems like nftables is not running (I'm not even sure the module can be loaded as it's inside a container)
```
mikrotik:/# nft list tables
netlink: Error: cache initialization failed: Invalid argument
```

**Comment by RomRider** [comment]
I've tried with the `iptables-legacy` package (symlinked it to /sbin/iptables) and it also doesn't work.

There's probably no way to fix this, according to https://forum.mikrotik.com/viewtopic.php?t=202838

**Comment by frankedwards** [comment]
Thanks, that's interesting.

I do see in a log entry
```
12:50:00 container,info,debug 2025/01/30 02:50:00 envknob: TS_DEBUG_FIREWALL_MODE="nftables"
```
... but not the other two entries per https://tailscale.com/kb/1294/firewall-mode

Also exit node also seems to be working, however I'm running the container on a CHR - I will re-test on Mikrotik hardware.

**Comment by RomRider** [comment]
I don't think mikrotik supports nftables (running a CCR2009 here).

This seems to be working better https://github.com/tailscale/tailscale/issues/10540#issuecomment-1908922102
But mikrotik doesn't support packet marking from within a container (stated on the link in my previous post)

**Comment by frankedwards** [comment]
Are you able to share your Dockerfile snippet... result for me
```
mikrotik:/# iptables -L -vvv
iptables v1.8.10 (legacy): can't initialize iptables table `filter': Module is wrong version
Perhaps iptables or your kernel needs to be upgraded.
```

**Comment by RomRider** [comment]
Sure (just the alpine part)
```Dockerfile
FROM alpine:3.19

RUN apk add --no-cache ca-certificates iptables iptables-legacy iproute2 bash openssh curl jq

RUN rm /sbin/iptables && ln -s /sbin/iptables-legacy /sbin/iptables
RUN rm /sbin/ip6tables && ln -s /sbin/ip6tables-legacy /sbin/ip6tables

RUN ssh-keygen -f /etc/ssh/ssh_host_rsa_key -N '' -t rsa
RUN ssh-keygen -f /etc/ssh/ssh_host_dsa_key -N '' -t dsa

COPY --from=build-env /go/bin/* /usr/local/bin/
COPY sshd_config /etc/ssh/
COPY tailscale.sh /usr/local/bin

EXPOSE 22
CMD ["/usr/local/bin/tailscale.sh"]
```

Also, I've removed `TS_DEBUG_FIREWALL_MODE=nftables` from the startup script as nftable is not available on mikrotik

**Comment by RomRider** [comment]
Some more findings... setting `--netfilter-mode=off` in `TAILSCALE_ARGS` makes the error go away (because it doesn't use iptables anymore) but I have still to understand the consequences of that... Everything still works for me with this setup (exit-node and subnet routing)

**Comment by frankedwards** [comment]
Yes, agree that exit node and other features appear to be OK when setting ```--netfilter-mode=off```. 

I guess we can conclude that use of iptables is platform-dependent and can be controlled for tailscale using the ```netfilter-mode``` flag. I'll make a comment in the README regarding this.

**Comment by PatchouliTC** [comment]
so which version of image could be use?
i'm using RB5009(ros version is 7.19.6);
when i use 0.1.33,it give me <iptables v1.8.10 (nf_tables): Could not fetch rule set generation id: Invalid argument> error;if i change to latest 0.1.35,it give me [MARK --set-mark 0x40000/0xff0000 --wait]: exit status 1: iptables: Invalid argument],both of them seems ipv4forward not work [traceroute can see ts ip next route success to container,but i can't visit ts device ip in lan but can visit it in container shell],so anyone can give a useable image or any tips to build iptable usful？
